Proposal: split geoService !important

AleksueiR opened the pull request:
Causes __Panic__ for 60min.

New services created: `layerRegistry`, `mapService` and `gapiService`. They do not directly depend on `geoService`.

`layerRegistry` manipulation with layers mostly (generate, register, remove, etc.)
`mapService` used to register map node and hold reference to the map object
`gapiService` holds reference to gapi

`gapi` loading is hoisted and now starts when the lib package is download, even before `app-seed` initializes app instances.

`identifyService` now accesses layers directly from `layerRegistry`, not storing them itself. This solves the growing number of layers when swapping configs. #367

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/381)
<!-- Reviewable:end -->

spencerwahl commented:
Reviewed 17 of 17 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381)*
<!-- Sent from Reviewable.io -->

alyec commented:
In javascript `!important` means not important.

---

Reviewed 17 of 17 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[src/app/geo/gapi.service.js, line 36 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQWHk-eGJ7F4IyUdDJ:-KCQWHk-eGJ7F4IyUdDK:1088472356) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/gapi.service.js#L36)):*
Why not allow chaining onto the gapiPromise (same question as mapService)?

---

*[src/app/geo/geo.service.js, line 83 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQVXD2LL04AUU2Uq3L:-KCQVXD2LL04AUU2Uq3M:-1691193406) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/geo.service.js#L83)):*
What do you think about moving the map creation into mapService?

---

*[src/app/geo/geo.service.js, line 124 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQUrFAmSrryJw8JbSX:-KCQUrFBkglSIiUXRJFj:-1086628870) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/geo.service.js#L124)):*
Is `data => resolve(data)` really necessary?

---

*[src/app/geo/identify.service.js, line 25 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQRh-Rvxu8tAJ3AI_A:-KCQRh-Rvxu8tAJ3AI_B:1174576876) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/identify.service.js#L25)):*
We probably need some sort of event / signaling system.  Without promise chains / some other initialization dependency tracking I think this could become fragile.

---

*[src/app/geo/identify.service.js, line 33 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQVuy_lGKJfiBBYAWX:-KCQVuy_lGKJfiBBYAWY:1378035889) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/identify.service.js#L33)):*
typo: layers

---

*[src/app/geo/identify.service.js, line 39 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQTJlDXe1jqWvSdWg_:-KCQTJlDXe1jqWvSdWga:-107924242) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/identify.service.js#L39)):*
Pretty experimental, but how about `Object.entries` instead of having to go with `keys().map(key => object)`?  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries

---

*[src/app/geo/identify.service.js, line 45 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQVswg8dgCqnV755ap:-KCQVswg8dgCqnV755aq:1378035889) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/identify.service.js#L45)):*
typo: layers

---

*[src/app/geo/init-map.directive.js, line 35 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQR73JAWf1wiZbPLNO:-KCQR73KsSly_Pqs7LL_:-476421981) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/ca95720fb8392acee94da35e9bc9afc9cc9107bb/src/app/geo/init-map.directive.js#L35)):*
One of the things I liked about this model was that the config was passed around as a plain object from here.  All of the geo modules could be tested directly without having to mock configService.  I'm not sure if there's a nice way to get that feature back without undoing a bunch of other work.

---

*[src/app/geo/layer-registry.service.js, line 40 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQPzynJcEaDOmcSHZk:-KCQPzynJcEaDOmcSHZl:2037768368) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/layer-registry.service.js#L40)):*
jsdoc missing

---

*[src/app/geo/layer-registry.service.js, line 73 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQQB24SwR7VOdEqqaS:-KCQQB24SwR7VOdEqqaT:-1922884546) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/layer-registry.service.js#L73)):*
is it a good idea for this to fail silently?

---

*[src/app/geo/layer-registry.spec.js, line 20 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQPgVb-DEEkmCqb_vF:-KCQPgVb-DEEkmCqb_vG:544218146) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/layer-registry.spec.js#L20)):*
this is probably a good time to const everything

---

*[src/app/geo/map.service.js, line 25 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQO_Z_RSSqn8es0N8-:-KCQO_ZaaZHRkRLVqVeU:2107743213) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/map.service.js#L25)):*
Would it make more sense for this to take mapNode as a promise?  The register could become a simple `.then` and you could avoid having to expose the promise internals in the `init` function.  Something like:

```js
isReady = $q.resolve(node).then(node => { service.mapNode = node; });
```

---

*[src/app/geo/map.service.js, line 50 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQNLEFfFzXOhnKsigx:-KCQNLEGMz2Ta3vVxWPz:1216177636) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/map.service.js#L50)):*
reject is not used

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-:-KCQKyEc_3CQhe0WnMQ9:-367810385)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[src/app/geo/gapi.service.js, line 36 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQWHk-eGJ7F4IyUdDJ:-KCQpQO3OuyKE8SsWgSQ:-53374812) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/gapi.service.js#L36)):*
I wanted to store `gapi` instance since it's truly immutable.

---

*[src/app/geo/geo.service.js, line 83 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQVXD2LL04AUU2Uq3L:-KCQY1hwYtKsaH2SplNG:-237418724) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/geo.service.js#L83)):*
Absolutely! I didn't have time to do that; adding comments to do that right now.

---

*[src/app/geo/geo.service.js, line 124 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQUrFAmSrryJw8JbSX:-KCQYFGI59s1B1xXStW9:-860517022) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/geo.service.js#L124)):*
Not sure; I'm not very familiar with what is going on here. @james-rae any thoughts?

---

*[src/app/geo/identify.service.js, line 33 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQVuy_lGKJfiBBYAWX:-KCQcGL5ngju5r4atSYr:-499146009) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/identify.service.js#L33)):*
Done.

---

*[src/app/geo/identify.service.js, line 39 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQTJlDXe1jqWvSdWg_:-KCQcIp2YenCRDKLZMXn:-1787386509) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/identify.service.js#L39)):*
I'll add `TODO` about this.

---

*[src/app/geo/identify.service.js, line 45 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQVswg8dgCqnV755ap:-KCQd7Ko70OXrKwqRbBt:-499146009) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/identify.service.js#L45)):*
Done.

---

*[src/app/geo/init-map.directive.js, line 35 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQR73JAWf1wiZbPLNO:-KCQdLuj23C5yWOZJ9-d:-750678053) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/ca95720fb8392acee94da35e9bc9afc9cc9107bb/src/app/geo/init-map.directive.js#L35)):*
Understood. For testing `geoService` I made a mock `configService`. We can probably move it into a separate file and share among geo test files. Would this be acceptable?

---

*[src/app/geo/layer-registry.service.js, line 40 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQPzynJcEaDOmcSHZk:-KCQpKNoF2Lqd9OJV5w4:-499146009) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/layer-registry.service.js#L40)):*
Done.

---

*[src/app/geo/layer-registry.service.js, line 73 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQQB24SwR7VOdEqqaS:-KCQg2KlYIr5VG7Hh1O1:-830342660) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/layer-registry.service.js#L73)):*
Probably not. Added a todo.

---

*[src/app/geo/layer-registry.spec.js, line 20 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQPgVb-DEEkmCqb_vF:-KCQiYvzz3D-rD0M4tGF:-499146009) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/layer-registry.spec.js#L20)):*
Done.

---

*[src/app/geo/map.service.js, line 25 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQO_Z_RSSqn8es0N8-:-KCQjAuNlH2LZwZleWc6:-1721955921) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/map.service.js#L25)):*
I would love that, but I don't understand something. If `mapNode` is a promise, who creates it? If it's created in `mapService`, how is it resolved by from outside by the `init-map` directive? If it's created by the `init-map` directive, then `mapNode` will be undefined when `isReady = $q.resolve(node).then(...)` is executed. 

---

*[src/app/geo/map.service.js, line 50 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQNLEFfFzXOhnKsigx:-KCQm3J4muQjHXhUZv8b:1725250986) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/map.service.js#L50)):*
Corrected.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381)*
<!-- Sent from Reviewable.io -->

spencerwahl commented:
Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/app/geo/identify.service.js, line 25 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQRh-Rvxu8tAJ3AI_A:-KCRZa43IV0wrQdPXKQD:-1437478186) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/72dbb8bc68a0d6c8274b42be58dec31e900acffa/src/app/geo/identify.service.js#L25)):*
Working on it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381)*
<!-- Sent from Reviewable.io -->

spencerwahl commented:
Reviewed 1 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Review status: 23 of 25 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/app/geo/geo.service.js, line 47 \[r3\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCWzDgmNXod7BRedP8I:-KCX2xEXVZN1EoklL6rB:-1572789896) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/93d701572eb11a1cffa539889b4c0bb385111565/src/app/geo/geo.service.js#L47)):*
No, `isReady` is only used initially to check if the dom node is registered by the `init-map` directive.

---

*[src/app/geo/identify.service.js, line 247 \[r3\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCX-NqHhHcqXiAkb2Et:-KCX440AOF6iE2-pjxdQ:-164627124) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/93d701572eb11a1cffa539889b4c0bb385111565/src/app/geo/identify.service.js#L247)):*
Sorry about that.

---

*[src/app/geo/init-map.directive.js, line 35 \[r1\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCQR73JAWf1wiZbPLNO:-KCX7Zy-C-3SObaxfjMI:625669301) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/ca95720fb8392acee94da35e9bc9afc9cc9107bb/src/app/geo/init-map.directive.js#L35)):*
This might work as well. Next PR?

---

*[src/app/geo/layer-registry.service.js, line 19 \[r3\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCX0cK0C19PXkIZE9x7:-KCX4vyFFnC3ojTWb1XI:-2085453783) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/93d701572eb11a1cffa539889b4c0bb385111565/src/app/geo/layer-registry.service.js#L19)):*
Good solution, I'll do that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381)*
<!-- Sent from Reviewable.io -->

spencerwahl commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 1 of 2 files at r4, 1 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/app/geo/geo.service.js, line 47 \[r3\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCWzDgmNXod7BRedP8I:-KCX9LIC4vzaWdrlENc8:-1411840808) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/93d701572eb11a1cffa539889b4c0bb385111565/src/app/geo/geo.service.js#L47)):*
Can we rename it so it's obvious / add some detail to the jsdoc (maybe split out the jsdoc into a separate entry for isReady)?

---

*[src/app/geo/identify.service.js, line 247 \[r3\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KCX-NqHhHcqXiAkb2Et:-KCXA53dGkbkn10RR9ig:1146509850) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/93d701572eb11a1cffa539889b4c0bb385111565/src/app/geo/identify.service.js#L247)):*
It was more of a general comment to see what people prefer.  But especially things like 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 1 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 6 of 6 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/app/geo/init-map.directive.js, line 27 \[r6\]](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381#-KC_xyq-rwZqnNdFHB-8:-KC_xyq-rwZqnNdFHB-9:-1970328045) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/8653ddb6d91ddf2c68bc2a68a516cd487604e946/src/app/geo/init-map.directive.js#L27)):*
typo: deregister

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381)*
<!-- Sent from Reviewable.io -->

ec-dev commented:
Reviewed 1 of 19 files at r3, 1 of 2 files at r4, 2 of 3 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381)*
<!-- Sent from Reviewable.io -->

spencerwahl commented:
Reviewed 3 of 6 files at r6, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/381)*
<!-- Sent from Reviewable.io -->
